### ISSUE TITLE
Improve dart:mirrors based on reflactable package

### ISSUE THREAD

**Issue by stevenroose** [dart-lang/sdk#26475]
The `reflectable` package fills in where `dart:mirrors` falls short. However, the default way to do reflection should still be by using a built-in package than a third-party one, right?

`reflectable` uses capabilities to indicate what kinds of reflection is required. I think it can be a good idea to also introduce capabilities to mirrors (f.e. `MirrorCapability` subclasses listed in the `@MirrorsUsed` annotation).

`reflectable` still has a lot of [limitations](https://github.com/dart-lang/reflectable/tree/master/reflectable#known-limitations), but I think some aspects can already be considered for inclusion in `dart:mirrors`.

I'm sure @eernstg can help.

**Comment by eernstg** [comment]
TL;DR  Maybe you could use the output from the transformer?  RD;LT

This issue is quite broad, but I'll outline my take on the overall situation and end with a concrete idea.

You could claim that a 'dart:' library is "more reliable" (in the sense that it will be used very heavily and debugged intensively, it will be maintained forever, etc.), but that would depend on the organizational context and the human beings involved, not on the technical details. So any frequently used 'package:' library should per se fit the bill just as well a 'dart:' library.

But, as you mention, reflectable has some limitations.

(Aside: If some of these limitations of reflectable are creating serious problems for you then it would be helpful to have an issue on https://github.com/dart-lang/reflectable/issues which explains the concrete situation. For instance, if we are going to consider adding a primitive for violating privacy protection then we have to know that it matters in real life.)

There are some limitations which are quite difficult to overcome (like privacy ;-), and trade-offs like "we want primitive X" and "any implementation with primitive X that we can envision will be slow" will have to be considered carefully. So extensions may or may not happen.

However, there is no big difference between having a core mirror package like 'dart:mirrors' and handling the same task in a 'package:' library (that is, something completely written in Dart), because the latter may rely on a carefully selected set of low-level primitives. I actually think the latter approach, primitives + normal Dart code, is cleaner and more flexible than 'dart:monolithic_magic_thing'.

From that point of view it might be better to fill in the missing bits in reflectable than starting a major rewrite of 'dart:mirrors' (where the techniques used by reflectable may not fit well at all)...

Focusing on the case where the task is covered by reflectable of today, and the situation where the use of a transformer is prohibitively inconvenient, there is one obvious step associated with `@MirrorsUsed` that could be taken:

Running in _non-transformed_ mode, reflectable is basically a wrapper around 'dart:mirrors' enhanced with capability checks. This is in a sense already what you are asking for, except that it does not use the capabilities to save space by selectively omitting mirror information.

However, you could run the transformer with '--mode=debug' such that the generated Dart code is available (only whenever something "significant" has been changed), and then use the generated code to create a `@MirrorsUsed` expression: The classes needed are the ones for which we have generated a `..ClassMirrorImpl`, and the symbols needed are the ones for which we have a 'VariableMirrorImpl`, `MethodMirrorImpl`, `ParameterMirrorImpl`, etc. With that `@MirrorsUsed` expression in place, you might get the effect you are asking for.

**Comment by stevenroose** [comment]
First of all, thanks for your detailed opinion. Let me formulate my opinion.

- First of all, I believe that reflection should be something the language provides, not something a third party package provides (not even if both are maintained by the same party). (Think Java's `java.lang.reflect` or Go's `import "reflect"`.)
Reflection is inherent to the language and the version of the language/SDK. When a new Dart version lands, the included `dart:mirrors` should be immediately compatible with the changes and developers should not have to wait for a parallel (but perhaps delayed) update to a third party package. 

- Second, I think considering Reflectable as "Dart reflection 2.0" is not correct. Reflectable is "An alternative approach to reflection by using code transformers, primarily aimed to minimize dart2js output". This is excellent. A lot of developers are obviously relieved to have a way to minimize their dart2js output while still being able to use reflection. However, this is a specific solution for some use cases and cannot substitute reflection in general.

- Lastly, and this is where my main concern surfaces, the decision of how to use reflection should be for the end user and not the library developers. When creating a library that needs some sort of reflection, you should be able to simply use the "default" method for reflection. The decision whether or not to use code transformers or some other kind of optimization should be up to the users of the library; in the end, they know what platform the code will run on. 

So in this regard, the ideal role of Reflectable would be to allow a developer, at build/deploy time, to replace all reflection functionality with code transformers. This is basically what it does now, but without requiring other packages to depend on Reflectable. So this would mean that all other packages are developed using `dart:mirrors` and that the Reflectable code transformer transforms these calls to Reflectable calls. 
Ironically, this is exactly what dart2js currently does. Inefficiently. All Reflectable adds is means to indicate what reflection functionality is actually required, but this is exactly what `@MirrorsUsed` is intended for. Sadly, `@MirrorsUsed` is still marked experiment and it's functionality is limited. However, the capability system used by Reflectable could in fact be integrated in `@MirrorsUsed`. 

To sum up:
There are many use cases of reflection that should not necessarily increase dart2js output drastically. (f.e. [method delegation](https://github.com/stevenroose/dart-cryptoutils/blob/02393eb2530d2f90b0e2709be4fa8cfa16134c1f/lib/src/hash/unmodifiable_uint8list.dart#L26) if `@MirrorsUsed` would support listing individual classes for which reflection is required) However [some platforms](https://github.com/flutter/flutter/issues/1150) don't support `dart:mirrors` because they are scared of the code compilation output size. So the solution is to use reflectable in libraries ([eg1](https://github.com/PointyCastle/pointycastle/commit/2da75e5a8d7bdbf95d08329add9f13b9070b75d4), [eg2](https://github.com/stevenroose/dart-websockets/commit/eac30eabe16f9749065b3a6030cca60695aaaec9)). However currently, when using Reflectable in a library, you require the users of that library, directly or indirectly, to [add an explicit dependency on Reflectable](https://github.com/dart-lang/reflectable/issues/84). This should be avoidable.
